### PR TITLE
Allow the use of a list comprehension to initialise an array

### LINKED ISSUE #2080: Allow the use of a list comprehension to initialise an array
### Describe the bug
A list comprehension must be saved to a variable in order to operate on the function, however currently it must be saved to a list variable. This is quite constraining. It would be nice to lift this restriction for NumPy arrays

### To Reproduce
Provide code to reproduce the behaviour:
```python
a = np.array([i*2 for i in range(5)])
```

### Error details
Provide the generated code, or the error message:
```
pyccel:
 |error [syntactic]: tmp.py [2,13]| The result of a list comprehension expression must be saved in a variable ([i * 2 for i in range(5)])
```

### Expected behavior
The error should still appear if the comprehension is not saved to a variable, but an exception should be added to allow the variable to be an array.

### Language
Syntactic error affects all languages

### Additional context
Needed by PyGyro

### PR DESCRIPTION
**Pull request by EmilyBourne**
Allow the use of a list comprehension to initialise an array. Fixes #2080 .
The type of the target that is created from the `FunctionalFor` is saved into the `FunctionalFor`. This extra parameter will also be useful when support for set comprehension and/or dictionary comprehension is added.

### CODE REVIEW

File: pyccel/parser/semantic.py [hunk 1]
```diff
@@ -3777,7 +3778,35 @@ def _visit_FunctionalFor(self, expr):
...
+        target_type_name = expr.target_type
+        if isinstance(target_type_name, DottedName):
+            lhs = target_type_name.name[0] if len(target_type_name.name) == 2 \
+                    else DottedName(*target_type_name.name[:-1])
+            first = self._visit(lhs)
+            if isinstance(first, Module):
+                conversion_func = first[target_type_name.name[-1]]
+            else:
+                conversion_func = None
+        else:
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/parser/semantic.py:3790]: /bot accept
Error handling

**Review by jalalium — APPROVED** [review]
Well done!

**Pull request merged**